iothub/service/src/Authentication/IotHubConnectionStringParser.cs

```diff
@@ -0,0 +1,103 @@
+    {
+        internal static IotHubConnectionString Parse(string iotHubConnectionString)
+        {
+            IDictionary<string, string> map = ToDictionary(iotHubConnectionString, IotHubConnectionStringConstants.ValuePairDelimiter, IotHubConnectionStringConstants.ValuePairSeparator);
+
+            string hostName = GetConnectionStringValue(map, IotHubConnectionStringConstants.HostNamePropertyName);
+            string sharedAccessKeyName = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessKeyNamePropertyName);
+            string sharedAccessKey = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessKeyPropertyName);
+            string sharedAccessSignature = GetConnectionStringOptionalValue(map, IotHubConnectionStringConstants.SharedAccessSignaturePropertyName);
+
+            Validate(hostName, sharedAccessKeyName, sharedAccessKey, sharedAccessSignature);
+
+            return new IotHubConnectionString(hostName, sharedAccessKeyName, sharedAccessKey, sharedAccessSignature);
+        }
+
+        private static void Validate(string hostName, string sharedAccessKeyName, string sharedAccessKey, string sharedAccessSignature)
+        {
+            string iotHubName = GetIotHubName(hostName);
+            if (string.IsNullOrWhiteSpace(sharedAccessKeyName))
+            {
+                throw new FormatException("Should specify SharedAccessKeyName.");
+            }
+
+            if (!(string.IsNullOrWhiteSpace(sharedAccessKey) ^ string.IsNullOrWhiteSpace(sharedAccessSignature)))
+            {
+                throw new FormatException("Should specify either SharedAccessKey or SharedAccessSignature.");
+            }
+
+            if (string.IsNullOrWhiteSpace(iotHubName))
+            {
+                throw new FormatException("Missing IoT Hub name.");
+            }
+
+            if (!string.IsNullOrWhiteSpace(sharedAccessKey))
+            {
+                // The conversion is to validate the provided value.
+                Convert.FromBase64String(sharedAccessKey);
+            }
+
+            if (SharedAccessSignatureParser.IsSharedAccessSignature(sharedAccessSignature))
+            {
+                SharedAccessSignatureParser.Parse(iotHubName, sharedAccessSignature);
+            }
+        }
+
+        private static string GetIotHubName(string hostName)
+        {
+            int index = hostName.IndexOf(IotHubConnectionStringConstants.HostNameSeparator, StringComparison.OrdinalIgnoreCase);
+            string iotHubName = index >= 0
+                ? hostName.Substring(0, index)
+                : hostName;
+            return iotHubName;
+        }
+
+        private static IDictionary<string, string> ToDictionary(string valuePairString, char kvpDelimiter, char kvpSeparator)
+        {
+            if (string.IsNullOrWhiteSpace(valuePairString))
+            {
+                throw new ArgumentException("Malformed token.");
+            }
```

Comment:
This private method is being used after validation that the argument (valuePairString) is not null. 
We have this check again to make sure any future reference to this method still validates it.

Not sure if the message in the exception is appropriate. Thoughts?